pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -0,0 +1,154 @@
+// limitations under the License.
+
+package eventstore
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"github.com/pipe-cd/pipe/pkg/app/api/service/pipedservice"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+// Getter helps get an event. All objects returned here must be treated as read-only.
+type Getter interface {
+	// GetLatest returns the latest event that meets the given conditions.
+	GetLatest(ctx context.Context, name string, labels map[string]string) (*model.Event, bool)
+}
+
+type Store interface {
+	// Run starts syncing the event list with the control-plane.
+	Run(ctx context.Context) error
+	// Getter returns a getter for retrieving an event.
+	Getter() Getter
+}
+
+type apiClient interface {
+	GetLatestEvent(ctx context.Context, req *pipedservice.GetLatestEventRequest, opts ...grpc.CallOption) (*pipedservice.GetLatestEventResponse, error)
+	ListEvents(ctx context.Context, req *pipedservice.ListEventsRequest, opts ...grpc.CallOption) (*pipedservice.ListEventsResponse, error)
+}
+
+type store struct {
+	apiClient    apiClient
+	syncInterval time.Duration
+	gracePeriod  time.Duration
+	logger       *zap.Logger
+
+	// Mark that it has handled all events that was created before this value.
+	milestone int64
+	// The key is supposed to be event-definition ID, a string consists of name and labels.
+	// And the value is the address to the latest Event.
+	latestEventMap atomic.Value
+}
+
+const (
+	defaultSyncInterval = time.Minute
+)
+
+// NewStore creates a new event store instance.
+// This syncs with the control plane to keep the list of events for this runner up-to-date.
+func NewStore(apiClient apiClient, gracePeriod time.Duration, logger *zap.Logger) Store {
+	s := &store{
+		apiClient:    apiClient,
+		syncInterval: defaultSyncInterval,
+		gracePeriod:  gracePeriod,
+		logger:       logger.Named("event-store"),
+	}
+	s.latestEventMap.Store(make(map[string]*model.Event))
+	return s
+}
+
+// Run starts runner that periodically makes the Events in the cache up-to-date
+// by fetching from the control-plane.
+func (s *store) Run(ctx context.Context) error {
+	s.logger.Info("start running event store")
+
+	syncTicker := time.NewTicker(s.syncInterval)
+	defer syncTicker.Stop()
+
+	// Do first sync without waiting the first ticker.
+	s.milestone = time.Now().Add(-time.Hour).Unix()
+	s.sync(ctx)
+
+	for {
+		select {
+		case <-syncTicker.C:
+			s.sync(ctx)
+
+		case <-ctx.Done():
+			s.logger.Info("event store has been stopped")
+			return nil
+		}
+	}
+}
+
+// sync fetches a list of events inside the range between own milestone and the current local time.
+// Only this function takes responsibility for updating the cache.
+func (s *store) sync(ctx context.Context) error {
+	// TODO: Use UTC and let control-plane convert it into the control-plane's Local time
+	// Unexpected behavior can be happened if Piped uses different timezone from the control-plane.
```

Comment:
I will look into this concern when working on another PR.

Reply:
I think the timezone difference is not a problem, because we are using Unix time (the number of seconds elapsed since January 1, 1970 UTC.).
But the `clock drift` may cause the problem. But that is often rare.

Btw, maybe we can leave the `to` parameter as zero to use `time.Now` of the server.

Reply:
>I think the timezone difference is not a problem, because we are using Unix time (the number of seconds elapsed since January 1, 1970 UTC.).

I put off thinking about it but after thinking, it's a basic thing and no problem as you mentiond.

Reply:
>Btw, maybe we can leave the to parameter as zero to use time.Now of the server.

Cool, I'll try it.

Reply:
>Btw, maybe we can leave the to parameter as zero to use time.Now of the server.

After thinking, it makes derivating the next's `from` tougher. As you mentioned, clock drifts is reraly happened. I guess it's enough to be as-is.

Reply:
Not tougher, I came up with a better solution just now.